servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyPipelinedConnection.java

```diff
@@ -0,0 +1,628 @@
+                }
+            };
+        }
+    }
+
+    /**
+     * Logically equivalent to {@link Completable#afterFinally(Runnable)} but relies upon internal queue CAS operations
+     * to prevent multiple executions (e.g. reduces a CAS operation).
+     */
+    private final class WritePopNextOperator implements CompletableOperator {
+        private final Node writeNode;
+
+        WritePopNextOperator(final Node writeNode) {
+            this.writeNode = writeNode;
+        }
+
+        @Override
+        public CompletableSource.Subscriber apply(final CompletableSource.Subscriber subscriber) {
+            return new CompletableSource.Subscriber() {
+                @Override
+                public void onSubscribe(final Cancellable cancellable) {
+                    subscriber.onSubscribe(() -> {
+                        try {
+                            cancellable.cancel();
+                        } finally {
+                            pollNext();
+                        }
+                    });
+                }
+
+                @Override
+                public void onComplete() {
+                    try {
+                        subscriber.onComplete();
+                    } finally {
+                        pollNext();
+                    }
+                }
+
+                @Override
+                public void onError(final Throwable t) {
+                    try {
+                        subscriber.onError(t);
+                    } finally {
+                        pollNext();
+                    }
+                }
+            };
+        }
+
+        private void pollNext() {
+            queuePop(writeQueueTailUpdater, writeNode);
+        }
+    }
+
+    private static final class Node {
+        private static final Node EMPTY_NODE = new Node();
+        private static final AtomicReferenceFieldUpdater<Node, Node> nextUpdater =
+                newUpdater(Node.class, Node.class, "next");
+        @Nullable
+        private volatile Node next;
+        final Runnable delayedSource;
+
+        Node(Runnable delayedSource) {
+            this.delayedSource = requireNonNull(delayedSource);
+        }
+
+        private Node() {
+            this.next = this;
+            this.delayedSource = () -> { };
+        }
+
+        boolean append(Node next) {
+            return nextUpdater.compareAndSet(this, null, next);
+        }
+
+        @Nullable
+        Node pop() {
+            return nextUpdater.getAndSet(this, EMPTY_NODE);
+        }
+
+        boolean isPopped() {
+            return next == EMPTY_NODE;
+        }
+
+        Node iterateToTail() {
+            Node prev = this;
+            Node next = prev.next;
+            while (next != null) {
+                prev = next;
+                next = next.next;
+            }
+            return prev;
+        }
+    }
+
+    private static final class DelayedSubscribeCompletable extends Completable {
```

Comment:
we could consider moving this to `concurrent.internal` in the future, but for now I just left them internal here.